README.md


<div id="top"></div>
<!--
*** Thanks for checking out the Best-README-Template. If you have a suggestion
*** that would make this better, please fork the repo and create a pull request
*** or simply open an issue with the tag "enhancement".
*** Don't forget to give the project a star!
*** Thanks again! Now go create something AMAZING! :D
-->



<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->



<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="#">
    <img src="resources/auau_logo.png" alt="Logo" height="80">
  </a>

  <h3 align="center">(Au)dio (Au)gmentation = auau</h3>

  <p align="center">
    It's all a work in progress since 2022-01-20
    <br />
  </p>
</div>



<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>



<!-- ABOUT THE PROJECT -->
## About The Project

This library is mainly intended for experimenting with audio data augmentation.
Let's see what happen!

Simple tests on [Google Colab](https://colab.research.google.com/drive/1cGLf8GYRZ3oYWylMnLXIo5Muv7BaiUua?usp=sharing)

<p align="right">(<a href="#top">back to top</a>)</p>



### Built With

This library is built on top of the following open-source projects:

* [librosa](https://librosa.org/doc/latest/index.html)
* [soundfile](https://pysoundfile.readthedocs.io/en/latest/)
* [numpy](https://numpy.org/)
* [torchaudio](https://pytorch.org/audio/stable/index.html)
* [pytorch](https://pytorch.org/)

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- GETTING STARTED -->
## Getting Started

To get started using `auau` you must have a python version `>=3.6`. Hereafter some considerations:
- `noise/` folder contains some examples from UrbanSound8K. They can be used as noise signals.
- `outputs/` folder is just an example containing some augmented audios.
- **You DO NOT require those folders for running the audio augmentations.**

### Prerequisites

The library is built on top of other audio augmentation libraries:
- librosa
- torch-audiomentations
- torchaudio

### Installation

_Install library in one command_


1. Execute the following on your preferred environment
   ```sh
   pip install git+https://github.com/MorenoLaQuatra/auau.git
   ```
2. That's it!

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- USAGE EXAMPLES -->
## Usage

```python
from auau.augmenter import Augmenter
from auau.loader import Loader
import soundfile as sf
import librosa

au = Augmenter()
loader = Loader()

signal, sr = loader.load_file("resources/test_audio.wav")

# White noise
noise_factor = 0.25
augmented_signal = au.add_white_noise(signal, noise_factor=0.25)
loader.save_file(f"outputs/test_audio_noise-{noise_factor}.wav", augmented_signal, sr)

# Time Strech
time_stretch_rate = 1.4
augmented_signal = au.time_stretch(signal, time_stretch_rate=time_stretch_rate)
loader.save_file(f"outputs/test_audio_timestretch-{time_stretch_rate}.wav", augmented_signal, sr)

# Pitch Scale
num_semitones = 2
augmented_signal = au.pitch_scale(signal, sampling_rate=sr, num_semitones=num_semitones)
loader.save_file(f"outputs/test_audio_pitch-{num_semitones}.wav", augmented_signal, sr)

# Random Gain
augmented_signal = au.random_gain(signal, min_gain=1, max_gain=1.1)
loader.save_file(f"outputs/test_audio_random_gain.wav", augmented_signal, sr)

# Polarity
augmented_signal = au.invert_polarity(signal)
loader.save_file(f"outputs/test_audio_invert_polarity.wav", augmented_signal, sr)

# External noise
noise_signal, noise_sr = librosa.load("noise/7383-3-0-0.wav")
au.resample(noise_signal, source_sr = noise_sr, target_sr = sr)
augmented_signal = au.add_noise(signal, noise_signal, noise_factor=0.25)
loader.save_file(f"outputs/test_noise.wav", augmented_signal, sr)

# Multiple external noise
list_noise_signals = []
list_noise_factors = []

noise_signal, noise_sr = librosa.load("noise/7383-3-0-1.wav")
list_noise_signals.append(librosa.resample(noise_signal, noise_sr, sr))
list_noise_factors.append(0.1)

noise_signal, noise_sr = librosa.load("noise/14113-4-0-0.wav")
list_noise_signals.append(librosa.resample(noise_signal, noise_sr, sr))
list_noise_factors.append(0.15)

augmented_signal = au.add_multiple_noise(signal, list_noise_signals=list_noise_signals, list_noise_factors=list_noise_factors)
loader.save_file(f"outputs/test_multiple_noise.wav", augmented_signal, sr)


# Random noise from folder
augmented_signal = au.add_random_noise(signal, sr, noise_folder="./noise/", min_noise_factor=0.01, max_noise_factor=0.05)
loader.save_file(f"outputs/test_random_noise.wav", augmented_signal, sr)

# Low pass filter
augmented_signal = au.low_pass_filter(signal, sr)
loader.save_file(f"outputs/test_audio_lowpass.wav", augmented_signal, sr)

# High pass filter
augmented_signal = au.high_pass_filter(signal, sr)
loader.save_file(f"outputs/test_audio_highpass.wav", augmented_signal, sr)

```



<!-- ROADMAP -->
## Roadmap

Please propose something you want to see in this library

- [x] Initial release `0.0.1`!
- [x] Make it work on GPU (if available)
- [ ] Additional augmentations (e.g., distortion or voice morphing)
- [ ] Pipelines
- [ ] Implement real documentation (besides examples)
- [ ] Standard support for multichannel audio (mainly stereo)

See the [open issues](https://github.com/MorenoLaQuatra/auau/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- CONTRIBUTING -->
## Contributing

- **V0.0.2** - This library is unstable and intented just for experimentation purposes.

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- LICENSE -->
## License

Distributed under the GNU General Public License v3.0 License. See `LICENSE` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- CONTACT -->
## Contact

Project Link: [https://github.com/MorenoLaQuatra/auau](https://github.com/MorenoLaQuatra/auau)

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- ACKNOWLEDGMENTS -->
## Acknowledgments

* [Valerio Velardo](https://github.com/musikalkemist) - This library is mainly inspired by Valerio's work!
* [README template](https://github.com/othneildrew/Best-README-Template)
* [Logo image](https://commons.wikimedia.org/wiki/File:Gold.svg)

<p align="right">(<a href="#top">back to top</a>)</p>


CHANGES.md
Changelog
---------

0.0.2
~~~~~
* Introduced some augmentations, some of them are not able to run on GPU
* Processing return Tensors for further processing
* Tensors are moved to specified device
* Added noise folder with some samples from UrbanSound8K

0.0.1
~~~~~
* Initial release.



test.py
from auau.augmenter import Augmenter
from auau.loader import Loader
import soundfile as sf
import librosa

au = Augmenter()
loader = Loader()

signal, sr = loader.load_file("resources/test_audio.wav")

# White noise
noise_factor = 0.25
augmented_signal = au.add_white_noise(signal, noise_factor=0.25)
loader.save_file(f"outputs/test_audio_noise-{noise_factor}.wav", augmented_signal, sr)

# Time Strech
time_stretch_rate = 1.1
augmented_signal = au.time_stretch(signal, time_stretch_rate=time_stretch_rate)
loader.save_file(f"outputs/test_audio_timestretch-{time_stretch_rate}.wav", augmented_signal, sr)

# Pitch Scale
num_semitones = 5
augmented_signal = au.pitch_scale(signal, sr=sr, num_semitones=num_semitones)
loader.save_file(f"outputs/test_audio_pitch-{num_semitones}.wav", augmented_signal, sr)

# Random Gain
augmented_signal = au.random_gain(signal, min_gain=1, max_gain=1.1)
loader.save_file(f"outputs/test_audio_random_gain.wav", augmented_signal, sr)

# Polarity
augmented_signal = au.invert_polarity(signal)
loader.save_file(f"outputs/test_audio_invert_polarity.wav", augmented_signal, sr)

# External noise
noise_signal, noise_sr = librosa.load("noise/7383-3-0-0.wav")
au.resample(noise_signal, source_sr = noise_sr, target_sr = sr)
augmented_signal = au.add_noise(signal, noise_signal, noise_factor=0.25)
loader.save_file(f"outputs/test_noise.wav", augmented_signal, sr)

# Multiple external noise
list_noise_signals = []
list_noise_factors = []

noise_signal, noise_sr = librosa.load("noise/7383-3-0-1.wav")
list_noise_signals.append(librosa.resample(noise_signal, noise_sr, sr))
list_noise_factors.append(0.1)

noise_signal, noise_sr = librosa.load("noise/14113-4-0-0.wav")
list_noise_signals.append(librosa.resample(noise_signal, noise_sr, sr))
list_noise_factors.append(0.15)

augmented_signal = au.add_multiple_noise(signal, list_noise_signals=list_noise_signals, list_noise_factors=list_noise_factors)
loader.save_file(f"outputs/test_multiple_noise.wav", augmented_signal, sr)


# Random noise from folder
augmented_signal = au.add_random_noise(signal, sr, noise_folder="./noise/", min_noise_factor=0.01, max_noise_factor=0.05)
loader.save_file(f"outputs/test_random_noise.wav", augmented_signal, sr)

# Low pass filter
augmented_signal = au.low_pass_filter(signal, sr)
loader.save_file(f"outputs/test_audio_lowpass.wav", augmented_signal, sr)

# High pass filter
augmented_signal = au.high_pass_filter(signal, sr)
loader.save_file(f"outputs/test_audio_highpass.wav", augmented_signal, sr)

# Flanger
augmented_signal = au.flanger(signal, sr)
loader.save_file(f"outputs/test_audio_flanger.wav", augmented_signal, sr)

# Phase
augmented_signal = au.phaser(signal, sr)
loader.save_file(f"outputs/test_audio_phaser.wav", augmented_signal, sr)



test_helper.py
from auau.augmenter_helper import AugmenterHelper
from auau.loader import Loader

ah = AugmenterHelper(noise_folder="./noise/", device="cpu")
loader = Loader()

example_list = [
    {"array": loader.load_file("resources/common_voice_it_22274222.mp3")[0], "sr": loader.load_file("resources/common_voice_it_22274222.mp3")[1]},
    {"array": loader.load_file("resources/common_voice_it_25565079.mp3")[0], "sr": loader.load_file("resources/common_voice_it_25565079.mp3")[1]},
    {"array": loader.load_file("resources/common_voice_it_26632586.mp3")[0], "sr": loader.load_file("resources/common_voice_it_26632586.mp3")[1]},
    {"array": loader.load_file("resources/common_voice_it_19975804.mp3")[0], "sr": loader.load_file("resources/common_voice_it_19975804.mp3")[1]},
    {"array": loader.load_file("resources/common_voice_it_21886466.mp3")[0], "sr": loader.load_file("resources/common_voice_it_21886466.mp3")[1]},
]

augmented_list = ah.augment(example_list, signal_key="array", sr_key="sr")

for i, d in enumerate(augmented_list):
    augmentation_type = d["augmentation_type"]
    loader.save_file(f"outputs/example_{i}_{augmentation_type}.wav", d["array"], d["sr"])


auau/augmenter.py
import random
import librosa
import numpy as np
import soundfile as sf
import copy
import logging
import torch
import math
import os
from auau.loader import Loader
from torch_audiomentations import HighPassFilter, LowPassFilter, Compose
import torchaudio.functional as FAudio

class Augmenter:
    '''This is the docstring for this class'''

    def __init__(self, device=None):
        super(Augmenter, self).__init__()
        if device is None:
            self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        else:
            self.device = device

        self.loader = Loader()
        self.allowed_extensions = ["wav", "mp3", ]
        self.list_augmentations = []
    
    def _convert_to_numpy(self, signal):
        if torch.is_tensor(signal):
            return signal.numpy()
        else:
            return signal
    
    def _convert_to_tensor(self, signal):
        if not torch.is_tensor(signal):
            return torch.from_numpy(signal)
        else:
            return signal

    def _move_to_cpu(self, signal):
        if torch.is_tensor(signal):
            if signal.device != "cpu":
                return signal.to("cpu")
            else:
                return signal
        else:
            return self._convert_to_tensor(signal)

    def _move_to_device(self, signal):
        return signal.to(self.device)

    def _tensor_on_device(self, signal):
        signal = self._convert_to_tensor(signal)
        signal = self._move_to_device(signal)
        return signal

    def _array_on_cpu(self, signal):
        signal = self._move_to_cpu(signal)      
        signal = self._convert_to_numpy(signal)
        return signal

    def _check_and_convert_mono(self, signal):
        if len(signal.shape) > 1:
            return torch.mean(signal, dim=1)
        else:
            return signal

    def resample(self, signal, source_sr, target_sr):
        """ Resampling using librosa, on CPU
        """
        signal = self._array_on_cpu(signal)
        signal = self._check_and_convert_mono(signal)
        return librosa.resample(signal, source_sr, target_sr)

    def add_white_noise(self, signal, noise_factor=0.25):
        """ Adding white noise, processing on self.device
        """
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        noise_signal = torch.normal(0, signal.std(), signal.size()).to(self.device)
        augmented_signal = signal + noise_signal * noise_factor
        return augmented_signal

    def time_stretch(self, signal, time_stretch_rate=1.5):
        """Time stretching implemented with librosa, processing on CPU
        """
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        signal = self._array_on_cpu(signal)
        augmented_signal = librosa.effects.time_stretch(signal, time_stretch_rate)
        return self._tensor_on_device(augmented_signal)

    def pitch_scale(self, signal, sr, num_semitones=2):
        """Pitch scaling implemented with librosa, processing on CPU
        """
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        augmented_signal = FAudio.pitch_shift(signal, sr, num_semitones)
        return self._tensor_on_device(augmented_signal)

    def gain(self, signal, gain):
        """
        Multiply signal by gain. 
        gain can either be a constant value, or a tensor with the same shape as signal
        """
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        if torch.is_tensor(gain):
            gain = self._move_to_device(gain)
        augmented_signal = signal * gain
        return augmented_signal

    def random_gain(self, signal, min_gain=1, max_gain=1.2):
        """
        Apply a random, constant gain
        """
        gain_rate = np.random.uniform(min_gain, max_gain) # constant value, no need to port it to device
        return self.gain(signal, gain_rate)

    def invert_polarity(self, signal):
        """Invert polarity, processing on device
        """
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        return signal * -1

    def add_noise(self, signal, noise, noise_factor=0.25):
        """Add provided noise, processing on device
        """
       
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        noise  = self._tensor_on_device(noise)
        noise = self._check_and_convert_mono(noise)

        repeats = math.ceil(len(signal) / len(noise))
        noise_signal = copy.deepcopy(noise)
        for n in range(repeats-1):
            noise_signal = torch.cat((noise_signal, noise))
        noise_signal = noise_signal[:len(signal)]
        augmented_signal = signal + noise_signal * noise_factor

        return augmented_signal

    def add_multiple_noise(self, signal, list_noise_signals=None, list_noise_factors=None):
        #match the length of the original signal (noise is repeated)
        if list_noise_signals == None or list_noise_factors==None:
            logging.ERROR("add_multiple_noise: list_noise_signals or list_noise_factors is set to None.")
            return signal
        
        if (len(list_noise_signals) != len(list_noise_factors)):
            logging.ERROR(f"add_multiple_noise: both list_noise_signals and list_noise_factors must have the same length {len(list_noise_signals)}!={len(list_noise_factors)}")
            return signal

        augmented_signal = copy.deepcopy(signal)
        for noise_index, noise_signal in enumerate(list_noise_signals):
            augmented_signal = self.add_noise(augmented_signal, noise_signal, list_noise_factors[noise_index])
        return augmented_signal

    def add_random_noise(self, signal, sr, noise_folder, traverse_folder=True, min_noise_factor=0.05, max_noise_factor=0.15):
        files = librosa.util.find_files(noise_folder, recurse=traverse_folder)
        noise_file = random.choice(files)
        noise_signal, noise_sr = self.loader.load_file(noise_file)
        noise_signal = self.resample(noise_signal, noise_sr, sr)
        noise_signal = self._tensor_on_device(noise_signal)
        noise_signal = self._check_and_convert_mono(noise_signal)
        noise_factor = random.uniform(min_noise_factor, max_noise_factor)
        return self.add_noise(signal, noise_signal, noise_factor)

    '''
    From torchaudio.functional
    '''

    def low_pass_filter(self, signal, sr, cutoff_freq=1000):
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        return FAudio.lowpass_biquad(signal, sr, cutoff_freq=float(cutoff_freq))

    def high_pass_filter(self, signal, sr, cutoff_freq=1000):
        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        return FAudio.highpass_biquad(signal, sr, cutoff_freq=float(cutoff_freq))

    def flanger(self, 
        signal, 
        sr, 
        delay: float = 0.0, 
        depth: float = 2.0, 
        regen: float = 0.0, 
        width: float = 71.0, 
        speed: float = 0.5, 
        phase: float = 25.0, 
        modulation: str = 'sinusoidal', 
        interpolation: str = 'linear'):

        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        signal = torch.unsqueeze(signal, dim=0)
        augmented_signal = FAudio.flanger(signal, 
            sr, 
            delay=delay,
            depth=depth,
            regen=regen,
            width=width,
            speed=speed,
            phase=phase,
            modulation=modulation,
            interpolation=interpolation
        )
        return augmented_signal[0, :]

    def phaser(self,
        signal, 
        sr,
        gain_in: float = 0.4, 
        gain_out: float = 0.74, 
        delay_ms: float = 3.0, 
        decay: float = 0.4, 
        mod_speed: float = 0.5, 
        sinusoidal: bool = True
        ):

        signal = self._tensor_on_device(signal)
        signal = self._check_and_convert_mono(signal)
        signal = torch.unsqueeze(signal, dim=0)
        augmented_signal = FAudio.phaser(signal, 
            sr,
            gain_in=gain_in,
            gain_out=gain_out,
            delay_ms=delay_ms,
            decay=decay,
            mod_speed=mod_speed,
            sinusoidal=sinusoidal
        )
        return augmented_signal[0, :]




auau/augmenter_helper.py

from auau.augmenter import Augmenter
import numpy as np
import collections
import random
import logging

class AugmenterHelper:
    '''This is the docstring for this class'''

    def __init__(self, noise_folder, device=None):
        super(AugmenterHelper, self).__init__()
        self.augmenter = Augmenter(device=device)
        self.noise_folder = noise_folder

    '''
    Robust Speech Recognition challenge
    0.15 -- Low pass 1000-1500 - apply
    0.15 -- High pass 500-1000 - apply
    0.45 -- Noise (random 8K)  - apply
    0.1 -- noise + LP
    0.1 -- noise + HP
    0.05 -- pitch shifting [-2;+2] random


    How to remove hardcoded?
    '''

    def augment(self, 
        list_dicts: list=[],
        signal_key: str="array",
        sr_key: str="sr"):
        
        list_augmentations = ["LP", "HP", "noise", "noise+HP", "noise+LP", "pitch"]
        list_probabilities = [0.15, 0.15, 0.45, 0.1, 0.1, 0.05]
        
        for i, d in enumerate(list_dicts):
            signal = d[signal_key]
            sr = d[sr_key]
            c = np.random.choice(list_augmentations, p=list_probabilities)

            if c == "LP":
                # range 1000-1500
                augmented_signal = self.augmenter.low_pass_filter(signal, sr, cutoff_freq=random.randint(1000,1200))
            elif c == "HP":
                # range 500-1000
                augmented_signal = self.augmenter.high_pass_filter(signal, sr, cutoff_freq=random.randint(800,1000))
            elif c == "noise":
                augmented_signal = self.augmenter.add_random_noise(signal, sr, noise_folder=self.noise_folder, max_noise_factor=0.1)
            elif c == "noise+LP":
                augmented_signal = self.augmenter.add_random_noise(signal, sr, noise_folder=self.noise_folder, max_noise_factor=0.1)
                augmented_signal = self.augmenter.low_pass_filter(augmented_signal, sr, cutoff_freq=random.randint(1000,1200))
            elif c == "noise+HP":
                augmented_signal = self.augmenter.add_random_noise(signal, sr, noise_folder=self.noise_folder, max_noise_factor=0.1)
                augmented_signal = self.augmenter.high_pass_filter(augmented_signal, sr, cutoff_freq=random.randint(800,1000))
            elif c == "pitch":
                augmented_signal = self.augmenter.pitch_scale(signal, sr, num_semitones=random.randint(-2,2))
            else:
                logging.ERROR("Augmentation not found, return original signal")

            list_dicts[i][signal_key] = augmented_signal
            list_dicts[i]["augmentation_type"] = c

        return list_dicts



